.github/workflows/ci.yml

```diff
@@ -3,6 +3,18 @@ name: ci
 on: [push, pull_request]
 
 jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - name: "Checkout the source code"
+        uses: actions/checkout@v2
+
+      - name: "Install Python"
+        uses: actions/setup-python@v2
+
+      - name: "Run pre-commit checks"
+        uses: pre-commit/action@v2.0.3
```

Comment:
A dedicated GitHub action is used here but it’s possible to install `pre-commit` on a local machine and run checks before a commit or at will.

Reply:
Any reason to use `pre-commit` directly? Can we instead run all the style/static checks directly within the job (as we did earlier for flake8)?
It has the benefit that people can easily run the checks locally without having to install an additional tool.

+1 for splitting out static analysis and tests to separate jobs though.

Reply:
* It serves as an umbrella for such tools. Basically, each tool defines a hook so each consumer doesn’t need to know about invocation details.
* It introduces a clear statement about pre-commit checks (hence the name). This repo doesn’t have `flake8` mentions outside CI so ATM local runs feel a bit like a tribal knowledge.
* It reuses environments / tools across repos since `pre-commit` installs its own versions.
* Since it uses a separate config, such tools do not clutter requirements. As you’ve noted, static analysis is different from tests.

I was under an impression that `pre-commit` is actually quite popular among Python projects. For example, [Django uses it](https://github.com/django/django/blob/main/.pre-commit-config.yaml), [Flask](https://github.com/pallets/flask/blob/main/.pre-commit-config.yaml), [Superset](https://github.com/apache/superset/blob/master/.pre-commit-config.yaml), etc. Also, as an anecdotal evidence, we use it at Lyft across hundreds of repos (not exclusively Python ones).

Reply:
FYI it seems the action is already deprecated https://github.com/pre-commit/action#readme

Reply:
Yeah. IMHO the pre-commit thing looks suspect. The docs indicate that they will not charge for the hook, but https://github.com/marketplace/pre-commit-ci already has a pricing plan attached. This is a shame really.

Reply:
Let's do this the old fashioned way - pre-commit doesn't add much value IMO to be considered just yet.

Reply:
Looks better, it can now be run locally as well by anybody new by looking at the ci config.